ccm_web/client/src/pages/BulkSectionCreate.tsx

```diff
@@ -0,0 +1,534 @@
+  }
+
+  const handleParseSuccess = (sectionNames: string[]): void => {
+    setSectionNames(sectionNames)
+    setPageState({ state: BulkSectionCreatePageState.Confirm, schemaInvalidation: [], rowInvalidations: [] })
+  }
+
+  const parseUpload = (file: File|undefined): void => {
+    if (file === undefined) {
+      resetPageState()
+      return
+    }
+    file.text().then(t => {
+      let lines = t.split(/[\r\n]+/).map(line => { return line.trim() })
+      // An empty file will resultin 1 line
+      if (lines.length > 0 && lines[0].length === 0) {
+        lines = lines.slice(1)
+      }
+
+      const schemaInvalidations: SectionsSchemaInvalidation[] = []
+
+      const sectionNameHeaderValidator: SectionsSchemaValidator = new SectionNameHeaderValidator()
+      schemaInvalidations.push(...sectionNameHeaderValidator.validate(lines))
+
+      if (schemaInvalidations.length !== 0) {
+        handleSchemaError([{ error: 'Error processing file', type: InvalidationType.Error }])
+        return
+      }
+
+      if (hasHeader(lines)) {
+        lines = lines.slice(1)
+      }
+
+      const rowInvalidations: SectionsRowInvalidation[] = []
+
+      const duplicateNamesInFileValidator: SectionRowsValidator = new DuplicateSectionInFileSectionRowsValidator()
+      rowInvalidations.push(...duplicateNamesInFileValidator.validate(lines))
+
+      const duplicatesNamesInCanvasValidator: DuplicateExistingSectionRowsValidator = new DuplicateExistingSectionRowsValidator()
+      rowInvalidations.push(...duplicatesNamesInCanvasValidator.validate(lines))
+
+      if (rowInvalidations.length !== 0) {
+        handleRowLevelInvalidationError(rowInvalidations)
+        return
+      }
+
+      handleParseSuccess(lines)
+    }).catch(e => {
+      // TODO Not sure how to produce this error in real life
+      handleSchemaError([{ error: 'Error processing file', type: InvalidationType.Error }])
+    })
+  }
+
+  const renderUploadHeader = (): JSX.Element => {
+    return <div className={classes.uploadHeader}>
+      <Typography variant='h6'>Upload your CSV File</Typography>
+      <br/>
+      <Typography><strong>Requirement:</strong> Your file should include one section name per line</Typography>
+      <Typography>View a <Link href='#'>CSV File Example</Link></Typography>
+    </div>
+  }
+
+  const renderFileUpload = (): JSX.Element => {
+    return <div className={classes.uploadContainer}>
+      <Grid container>
+        <Grid item xs={12}>
+          <FileUpload onUploadComplete={uploadComplete}></FileUpload>
+        </Grid>
+      </Grid>
+      <Backdrop className={classes.backdrop} open={isExistingSectionsLoading}>
+        <Grid container>
+          <Grid item xs={12}>
+            <CircularProgress color="inherit" />
+          </Grid>
+          <Grid item xs={12}>
+          <Typography>Loading Section Information</Typography>
+          </Grid>
+        </Grid>
+      </Backdrop>
+    </div>
+  }
+
+  const renderUpload = (): JSX.Element => {
```

Comment:
Can these `render*` constants pulled out and reused based on the feature type. I do see then used between `Canvas Gradebook Converter` and `BulkCreateSection`. Like `renderRowLevelErrors`, `renderTryAgainButton`, `renderUploadAgainButton`, `renderTopLevelErrors`, `renderInvalidUpload` etc

I know some of the elements that are rendered is not same but can be generalized based on Feature type as input

```
renderTryAgainButton(feature){
if(feature === 'Gradebook'){
<Link> Upload </Link>
}else {
<Link> Upload </Link>
<p> something </p>
}
}
```

General purpose functions needs to be generalized. The reason this is important is we have more file upload stuff coming with `Add Users` , `3rd party Gradebook`. This process would be simp-lied if the code is extracted and made it as reusable components. Don't need to do this here but can be take as separate issue. I know you have created an [issue](https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/85) around reusing confirmation table. But I think I am asking something different.

Reply:
My preference is to build out the UX and then combine reusable code as I see a pattern where it provides clear benefit.  I get the feeling from this review that others want to optimize it right away.  I don't know that the pattern has really emerged after 2 cases but as always there is probably a more perfect way to implement it.

With regards to the above suggestion I definitely do not like the specific idea of introducing a check like if (feature==='Gradebook'), I know it's just an example but it makes the code more breakable in attempting to reduce similar code and I much prefer the other end of that spectrum.   I can take a crack at code sharing across components, but I would prefer to do it as it seems to naturally give me more direct benefit.   This could be part of the earlier issue, or a separate one, or maybe its just an _issue_ with me developing code lol

Reply:
I feel any time you see a code naming doing similar thing it is good to generalize it. 
I have seem people whining about it and I did not get an approval unless until I made those changes, even that mean it's a 8 hrs change. I always heard it is better to spend time now and reap the benefit later. LOL. I will leave it how you want to do as a separate issue or same issues as [here](https://github.com/tl-its-umich-edu/canvas-course-manager-next/issues/85).

This would make the file little shorter

Reply:
It would make two files shorter and one file longer (or existant) and it could end up working out perfectly for a new component, or force component to follow a design that doesn't fit quite right, or modifying the shared code to account for another use case, which is the concern to me if you are doing the optimizations early in the process.  Being forced to conform in a slightly contrived way isn't necessarily bad, though, as it could reduce the number of places the code can go wrong, or be maintained.  
I'm definitely for reusing code wherever it makes sense, but also like to be comfortable I understand what the common use case is going to be before I knock it out, if that makes sense.

Reply:
I would like to differentiate between re-using code that is more or less identical, and code that looks similar in terms of patterns but has some differences in logic (which might require function parameters when making them re-usable, or other non-simple patterns). For the latter, it makes some sense to me to wait to see where the similarities are across a couple of use cases. The things I've commented on in the PR seemed like the former: low-hanging fruit for extraction and re-use (styles and row cell), but of course I could be missing some piece of complexity.

It does seem like there are some differences of opinion here, so I will be flexible and adaptable to others' work styles. However, I do ask that when you are working, if you find yourself copying and pasting and not modifying blocks of code, give it a try to extract that. I think this will save you time later and lead to more maintainable code and a consistent user interface.

Reply:
@chrisrrowland what is your verdict making a template component catering to  `render` function in both section creation and Gradebook converter? 

Note: I am not asking for change but want to understand what's your thinking for future.

Reply:
He did start working on this. https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/84/commits/e2b72caad3611340cf3ede53641588113165ab2e